File: src/main/java/io/snyk/gradle/plugin/SnykTask.java
package io.snyk.gradle.plugin;

import org.gradle.api.DefaultTask;
import org.gradle.api.GradleException;
import org.gradle.api.logging.Logger;
import org.gradle.api.provider.Property;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.Optional;

public abstract class SnykTask extends DefaultTask {


    protected Logger log = getLogger();
    
    @Input
    @Optional
    protected abstract Property<String> getSnykArguments();
    
    @Input
    @Optional
    protected abstract Property<String> getSnykSeverity();
    
    @Input
    @Optional
    protected abstract Property<String> getSnykApi();

    protected Runner.Result runSnykCommand(String command) {
        if (getSnykArguments().getOrNull() != null) {
            command += " " + getSnykArguments().get();
        }

        if (getSnykSeverity().getOrNull() != null) {
            command += " --severity-threshold=" + getSnykSeverity().get();
        }
        return Runner.runSnyk(command);
    }

    protected void authentication() {
        log.debug("check auth");
        String envToken = System.getenv("SNYK_TOKEN");
        if (envToken != null && !envToken.isEmpty()) {
            log.debug("api token found in env");
            return;
        }

        String apiConfigToken = Runner.runSnyk("config get api").getOutput().trim();
        if (apiConfigToken != null && !apiConfigToken.isEmpty()) {
            log.debug("api token found");
            return;
        }

        if (getSnykApi().getOrNull() == null || getSnykApi().get().isEmpty()) {
            throw new GradleException("No API key found");
        }

        Runner.Result authResult = Runner.runSnyk("auth "+ getSnykApi().get());
        if (authResult.failed()) {
            throw new GradleException("snyk auth went wrong: " + authResult.getOutput());
        }
    }
}

File: src/main/kotlin/io/snyk/gradle/plugin/kt/SnykTask.kt
package io.snyk.gradle.plugin.kt

import java.io.ByteArrayOutputStream
import javax.inject.Inject
import org.gradle.api.DefaultTask
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.provider.ListProperty
import org.gradle.api.provider.MapProperty
import org.gradle.api.provider.Property
import org.gradle.api.provider.Provider
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.InputFile
import org.gradle.api.tasks.Optional
import org.gradle.api.tasks.TaskAction
import org.gradle.api.tasks.UntrackedTask
import org.gradle.process.ExecOperations

@Suppress("UnstableApiUsage") // UntrackedTask is incubating
@UntrackedTask(because = "The Snyk CLI should always execute")
abstract class SnykTask @Inject constructor(
  private val executor: ExecOperations
) : DefaultTask() {

  @get:Input
  abstract val snykToken: Property<String>

  @get:InputFile
  abstract val snykCli: RegularFileProperty

  @get:Input
  abstract val command: Property<String>

  @get:Input
  abstract val arguments: ListProperty<String>

  @get:Input
  abstract val environmentVariables: MapProperty<String, String>

  @get:Input
  @get:Optional
  abstract val severityThreshold: Property<SnykExtension.Severity?>

  @get:Input
  @get:Optional
  abstract val integrationName: Property<String?>

  @TaskAction
  fun executeSnykCommand() {
    val arguments: MutableList<String> = arguments.getOrElse(emptyList()).toMutableList()

    arguments.add(0, command.get())
    arguments.addArg("severity", severityThreshold.map { it.cliArg })
    arguments.addArg("integration-name", integrationName)

    val snykCli = snykCli.get()

    val (result, output) = ByteArrayOutputStream().use { output ->
      val result = executor.exec {
        executable = snykCli.asFile.canonicalPath

        args = arguments
        environment(environmentVariables.get())

        standardOutput = output
        errorOutput = output

        isIgnoreExitValue = true
      }
      result to output.toString().trim()
    }

    logger.quiet(
      """
        Executed Snyk, exit value:${result.exitValue}
        $output
      """.trimIndent()
    )
  }

  companion object {
    private fun MutableList<String>.addArg(
      argName: String,
      argValue: Provider<String?>,
    ) {

      val value = argValue.orNull
      if (value != null && this.none { arg -> "--$argName" in arg }) {
        this += "--$argName=${value}"
      }
    }
  }
}
